File: vorticity_solve.py
import numpy as np
import matplotlib.pyplot as plt

def vorticity_solve():
    # function parameters
    Nx, Ny = 256, 256
    dt, t_final = 1e-3, 100.0
    n_timesteps = int(np.floor(t_final/dt))

    # Grid spacing
    dx = 2.*np.pi/Nx
    dy = 2.*np.pi/Ny

    nu = 5.9e-30
    nu_damping = 0.05
    A0 = 1.0
    R = 0.9
    kmax, kmin = 9, 7

    # Hyperdiffusivity
    p_hyper = 8

    # Discretized grid
    xx = np.arange(0, Nx)*dx
    yy = np.arange(0, Ny)*dy

    w = initial_vorticity(Nx, Ny, dx, dy)

    # Matrices of wavenumbers
    kx = np.ones((1, Ny)).T * (np.arange(-Nx/2, Nx/2))
    ky = np.reshape(np.arange(-Ny/2, Ny/2), (1, Ny)).T * np.ones((1, Nx))

    # forcing
    k_mod = np.sqrt(kx**2 + ky**2)
    Mx_scale = (k_mod<=kmax) * (k_mod>=kmin)
    forcing_hat = Mx_scale * np.ones((Ny, Nx))
    A0_num = A0*Nx*Ny/(np.sqrt(np.pi*(kmax**2 - kmin**2)))*(np.sqrt(2.0)/1)

    kx = 1j*kx
    ky = 1j*ky

    # Spherical dealiasing
    dealias = (np.abs(kx) < (2.0/3.0)*(Nx/2.0)) * (np.abs(ky)<(2.0/3.0)*(Ny/2.0))
    # Laplacian in fourier space
    k2_visc = kx**2 + ky**2
    k2_poisson = k2_visc
    k2_poisson[int(Nx/2),int(Nx/2)] = 1
    k_hyperdiff = -((-1)**p_hyper)*k2_visc**p_hyper

    w_hat = np.fft.fftshift(np.fft.fft2(w))

    print("Entering time loop... \n")
    # Update vorticity through each time step
    for iteration_time in range(0, n_timesteps):
        if np.mod(iteration_time, 100)==0:            
            seconds = np.round(iteration_time*dt, 4)
            print("Time: {0}s".format(seconds))
            w = np.real(np.fft.ifft2(np.fft.ifftshift(w_hat)))
            plt.imshow(w.T, cmap='hot')
            #plt.pcolormesh(yy, xx, w, cmap="hot")
            if iteration_time==0:
                plt.colorbar()
            plt.title("2D Navier Stokes - Vorticity")
            #plt.pause(1e-7)
            plt.savefig("W_images/vorticity_{0}".format(int(iteration_time/100)))
            # Numerical check for imcompressibility
            """
            if np.mod(iteration_time, 100)==0:
            psi_hat = -w_hat/k2_poisson                 # Solve Poisson's eqn
            u       = np.real(np.fft.ifft2(np.fft.ifftshift( ky*psi_hat)))# u = D_y psi
            v       = np.real(np.fft.ifft2(np.fft.ifftshift(-kx*psi_hat)))# v =-D_x psi
            u_hat   = np.fft.fftshift(np.fft.fft2(u))
            v_hat   = np.fft.fftshift(np.fft.fft2(v))
            u_x     = np.real(np.fft.ifft2(np.fft.ifftshift( kx*u_hat )))
            v_y     = np.real(np.fft.ifft2(np.fft.ifftshift( ky*v_hat )))
            grad_U[int(iteration_time/100)]  = np.abs(u_x + v_y).sum()
            """
        w_hat, forcing_hat = update_vorticity(w_hat, Nx, Ny, dt, kx, ky,
            k2_poisson, k_hyperdiff, dealias, Mx_scale, A0_num, R, nu,
            nu_damping, forcing_hat)
    print("Simulation finished. Showing final plot...")
    # Plot final w
    plt.pcolormesh(yy, xx, w, cmap="hot")
    plt.title("2D Navier Stokes - Vorticity")
    plt.colorbar()
    plt.show()


def update_vorticity(w_hat, Nx, Ny, dt, kx, ky, k2_poisson, k_hyperdiff, dealias,
        Mx_scale, A0_num, R, nu, nu_damping, forcing_hat):
    """
    Poissons equation is solved in Fourier space.
    In Fourier space the derivatives are multiplications of k2
    We perform multiplications in real space and derivative in fourier space.
    Compute the streamfunction, get velocity and gradient of vorticity
    """
    psi_hat = -w_hat/k2_poisson                 # Solve Poisson's eqn
    u       = np.real(np.fft.ifft2(np.fft.ifftshift( ky*psi_hat)))# u = D_y psi
    v       = np.real(np.fft.ifft2(np.fft.ifftshift(-kx*psi_hat)))# v =-D_x psi
    w_x     = np.real(np.fft.ifft2(np.fft.ifftshift( kx*w_hat)))  # x derivative of vorticity
    w_y     = np.real(np.fft.ifft2(np.fft.ifftshift( ky*w_hat)))  # y derivative of vorticity

    conv     = np.add(np.multiply(u,w_x), np.multiply(v,w_y))
    conv_hat = np.fft.fftshift(np.fft.fft2(conv))
    conv_hat = conv_hat * dealias              # Spherical dealiasing

    #Mx_random       = np.random.rand(Ny, Nx)
    #random_part_hat = Mx_scale * np.exp(2*np.pi*1j*Mx_random)
    forcing_hat     = 0#np.sqrt(1-R*R)*A0_num*random_part_hat + R*forcing_hat
    damping_hat     = -nu_damping*w_hat

    # Solution at next step
    w_hat_new = ((1+0.5*dt*nu*k_hyperdiff)/(1-0.5*dt*nu*k_hyperdiff))*w_hat \
        + dt*((conv_hat+forcing_hat+damping_hat)/(1-0.5*dt*nu*k_hyperdiff))

    return w_hat_new, forcing_hat

def initial_vorticity(Nx, Ny, dx, dy):
    """
    Inputs : Number of grid points in the x and y directions (Nx, Ny) and
             the respective grid spcaings (dx, dy)

    Expressions for the vorticities can be changed to give a prefered initialization
    as long as the matrix sizes remain defined over the discretized grid, ie. defined
    by the ii and jj meshgrids.

    Returns : The vorticities w and theta defined over the discretized grid
              defined by the number of grid points (Nx, Ny) and spacing (dx, dy).
              Vorticities will have shape (Nx, Ny).
    """
    print("Generating initial fields...")
    # Defining the Gaussian parameters
    x01 = 0.75*np.pi
    x02 = (5.0/4)*np.pi
    x03 = (5.0/4)*np.pi

    y01 = np.pi
    y02 = np.pi
    y03 = np.pi + (np.pi/(2.0*np.sqrt(2.0)))

    r0  = (1.0/np.pi)**2
    # Meshgrids for the inintialization of the vorticity field
    jj, ii = np.meshgrid(np.arange(1, Nx+1), np.arange(1, Ny+1))
    print("w vorticity")
    w = np.exp(-((ii*dx-x01)**2+(jj*dy-y01)**2)/r0)     \
        + np.exp(-((ii*dx-x02)**2+(jj*dy-y02)**2)/r0)   \
        -0.5*np.exp(-((ii*dx-x03)**2+(jj*dy-y03)**2)/r0)
    w = w*(1.0/(np.pi*r0))
    # theta vorticity...

    print("Initialization complete.\n")
    return w


#def update_data_file(seconds, w):
#    with open('vorticity_data.csv', mode='w') as data_file:
#        vorticity_write = csv.writer(data_file, delimiter=',',
#            quotechar='"', quoting=csv.QUOTE_MINIMAL)
#        vorticity_write.writerow([seconds, w])


vorticity_solve()

File: vorticity_solve.jl
using FFTW
using Dierckx
using PyPlot

# Tidy up and comment function doc string
function vorticity_solve()
    # Function parameters (Move to fix_params?)
    nu_damping  = 0.05
    Nx, Ny      = 256, 256
    dt, t_final = 1e-3, 100.0
    n_timesteps = Int(floor(t_final/dt))

    # Grid spacing is multiple of 2pi for easy FT
    dx = 2*pi/Nx
    dy = 2*pi/Ny

    nu = 5.9e-30
    nuP = nu
    A0, R = 1, 0.9
    kmax, kmin = 9, 7

    # Hyperdiffusivity
    p_hyper = 8

    # Create the discretized grid
    xx = [0:Nx-1...]*dx
    yy = [0:Ny-1...]*dy

    w, theta = initial_vorticity(Nx, Ny, dx, dy)

    # Matrices of wavenumbers with shapes (Ny, Nx)
    kx = ones(1, Ny+1)' *([-Nx/2:Nx/2...])'
    ky = ([-Ny/2: Ny/2...]) * ones(1, Nx+1)

    # Forcing
    k_mod = sqrt.(kx.^2 .+ ky.^2)
    Mx_scale = (k_mod .< kmax) .* (k_mod .> kmin) # Filter
    forcing_hat = Mx_scale.*(ones(Ny,Nx))
    A0_num = A0*Nx*Ny/sqrt(pi*(kmax^2 - kmin^2))*(sqrt(2))

    kx = im*kx
    ky = im*ky
    # ∇^2 in Fourier space
    k2_visc = kx.^2 + ky.^2
    k_hyperdiff = -((-1.0)^p_hyper)*k2_visc.^p_hyper
    k2_poisson  = copy(k2_visc)
    k2_poisson[Int(Nx/2), Int(Nx/2)] = 1      # fixed Laplacian in Fourier space for Poisson

    dealias = (abs.(kx).<(2.0/3.0)*(Nx/2.0)) .* (abs.(ky).<(2.0/3.0)*(Ny/2.0))

    w_hat = fftshift(fft(w))
    println(sum(w_hat))
    theta_hat = fftshift(fft(theta))

    # Initial particle position
    x0, y0 = 0.5*Nx*dx, 0.5*Ny*dy

    println("Entering time loop... \n")
    # Update vorticity through each time step
    for iteration_time in 1:n_timesteps
        break
        w_hat, theta_hat, forcing_hat = update_vorticity!(w_hat, theta_hat, kx, ky, k2_poisson, k_hyperdiff, dealias, dt, Nx, Ny, Mx_scale, R, A0_num, nu, nuP, nu_damping, forcing_hat)
        #theta_hat = copy(theta_hat_new)
        #psi_hat = -w_hat./k2_poisson

        # Particle tracking
#        u = real.(ifft(ky.*psi_hat))
#        v = real.(ifft(-kx.*psi_hat))
#        x0, y0 = update_particle_state!(xx, yy, u, v, dt, x0, y0)

        if mod(iteration_time, 100)==0
            println("Time ", iteration_time*dt, "s  -  ", iteration_time/n_timesteps *100, "% ...")
            w = real.(ifft(ifftshift(w_hat)))
            pcolor(xx, yy, w)
            PyPlot.colorbar()
            title("2D Navier Stokes - Vorticity")
            PyPlot.pause(0.0005)
            PyPlot.clf()
        end
    end
    #PyPlot.show()
end

# Clean function parameters
function update_vorticity!(w_hat, theta_hat, kx, ky, k2_poisson, k_hyperdiff, dealias, dt, Nx, Ny, Mx_scale, R, A0_num, nu, nuP, nu_damping, forcing_hat)
    """
    Poissons equation is solved in Fourier space.
    In Fourier space the derivatives are multiplications of k2
    We perform multiplications in real space and derivative in fourier space.
    Compute the streamfunction, get velocity and gradient of vorticity
    """
    psi_hat = -w_hat./k2_poisson    # Solve Poissons eqn
    u   = real.(ifft(ifftshift( ky.*psi_hat))) # u = D_y psi
    v   = real.(ifft(ifftshift(-kx.*psi_hat))) # v = - D_x psi
    w_x = real.(ifft(ifftshift( kx.*w_hat)))   # x derivative of vorticity
    w_y = real.(ifft(ifftshift( ky.*w_hat)))   # y derivative of vorticity
    theta_x = real.(ifft(ifftshift( kx.*theta_hat))) # x derivative of vorticity
    theta_y = real.(ifft(ifftshift( ky.*theta_hat))) # y derivative of vorticity

    convect     = u.*w_x + v.*w_y
    convect_hat = fftshift(fft(convect))

    convect_theta     = u.*theta_x + v.*theta_y
    convect_theta_hat = fftshift(fft(convect_theta))

    # Spherical dealiasing
    convect_hat = dealias.*convect_hat
    convect_theta_hat = dealias.*convect_theta_hat
    Mx_random       = rand(Ny, Nx)
    random_part_hat = Mx_scale.*exp.(2.0*pi*im*Mx_random)
    forcing_hat     = sqrt(1.0-R*R)*A0_num*random_part_hat + R*forcing_hat
    damping_hat     = -nu_damping*w_hat

    # Solution at next step
    w_hat_new = ((1.0 .+(0.5)*dt*nu*k_hyperdiff)./(1.0 .-(0.5)*dt*nu*k_hyperdiff)).*w_hat + dt*((convect_hat+forcing_hat+damping_hat)./(1.0 .-(0.5)*dt*nu*k_hyperdiff))

    theta_hat_new = ((1.0 .+(0.5)*dt*nuP*k_hyperdiff)./(1.0 .-(0.5)*dt*nuP*k_hyperdiff)).*theta_hat + dt*((convect_theta_hat)./(1.0 .-(0.5)*dt*nuP*k_hyperdiff))

    return w_hat_new, theta_hat_new, forcing_hat

end

function update_particle_state!(xx, yy, u, v, dt, x0, y0)
    u0 = Spline2D(xx, yy, u)(x0, y0)
    v0 = Spline2D(xx, yy, v)(x0, y0)

    x0_new = x0+u0*dt
    y0_new = y0+v0*dt

    return x0_new, y0_new
end

function initial_vorticity(Nx, Ny, dx, dy)
    """
    Inputs : Number of grid points in the x and y directions (Nx, Ny) and
             the respective grid spcaings (dx, dy)

    Expressions for the vorticities can be changed to give a prefered initialization
    as long as the matrix sizes remain defined over the discretized grid.

    Returns : The vorticities w and theta defined over the discretized grid
              defined by the number of grid points (Nx, Ny) and spacing (dx, dy).
              Vorticities will have shape (Nx, Ny).
    """
    println("Generating initial fields...")
    xx = [0:Nx-1...]*dx
    yy = [0:Ny-1...]*dy
    # x0i and y0i for i in {1,2,3} are phase shift parameters
    x01 = (0.75)*pi
    x02 = (5.0/4.0)*pi
    x03 = x02

    y01 = pi
    y02 = pi
    y03 = y02+(pi/(2.0*sqrt(2.0)))

    r0 = (1.0/pi)^2
    # ii, jj are meshgrids to vectorise the vorticity field (w) initialization
    ii, jj = meshgrid([1:Nx...], [1:Ny...])
    println("w vorticity...")
    w = exp.(-((ii*dx .-x01).^2+(jj*dy .-y01).^2)/r0) .+ exp.(-((ii*dx .-x02).^2+(jj*dy .-y02).^2)/r0) .-
        0.5.*exp.(-((ii*dx .-x03).^2+(jj*dy .-y03).^2)/r0)
    w = w*(1.0/(pi*r0))
    println("θ vorticity...")
    theta = zeros(Nx, Ny)
    @inbounds for i in 1:Nx
        @simd for j in 1:Ny
            rad   = (xx[i]-pi).^2 + (yy[j]-pi).^2
            theta[i,j] = (3.0/4.0)*exp(-rad)
        end
    end
    println("Initialization complete. \n")
    return w, theta
end

function meshgrid(x, y)
    """
    Inputs : Two 1 dimensional arrays (x, y)

    Generates the meshgrid from input vectors x and y.

    Returns : Matrices X and Y with shapes (length(x), length(y)), where the
              X matrix has x repeated over its columns and Y matrix has y
              repeated over its rows.

    Example : x = [1 2 3], y = [4 5]
              Then meshgrid(x,y) returns
              X = [1 1; 2 2; 3 3]
              Y = [4 5; 4 5; 4 5]
    """
    nx = length(x)
    ny = length(y)
    X = zeros(nx,ny)
    Y = zeros(nx,ny)
    @inbounds for i in 1:nx
        @inbounds @simd for j in 1:ny
            X[i, j] = x[i]
            Y[i, j] = y[j]
        end
    end
    return X, Y
end

vorticity_solve()
